### PR TITLE
Type check for better fortran interface support

### PR DESCRIPTION
**Pull request by danbeibei**
This pull request addresses two current limitations:
- As derived type are accessed via an integer representing the instance address, all derived type are the same type from the wrapper point of view. Polymorphism with derived type is not possible.
- Arrays with non corresponding type/shape to the underlying Fortran subroutines are sometimes passed without raising a TypeError (I guess automatique conversion from f2py). Polymorphism with different numpy array dtype or shape is not always possible: issue #107 

This pr adds type and shape checks in the python wrapper function. It raises TypeError if the type/shape provided to the python function does not match the type/shape of the wrapped Fortran subroutine.
It allows better support for Fortran interfaces. For instance, it enables one to wrap:
- A Fortran interface with kind=4 and kind=8 implementations of a subroutine.
- A Fortran interface with 2d arrays and 3d arrays implementations of a subroutine.
- A Fortran interface with implementations for two different derived types.

Known limitation: as Fortran `logical` is mapped to C `int` there is no differentiation of python `bool` and `int32`.

It should solve issue #107 

A command line flag is added to activate those checks, so that old behavior is still possible and is the default.
The flag could be removed so that the checks are always added, but this could break existing code. It does break some tests in the exemple directory, though they seem easily fixable.

### CODE REVIEW

File: f90wrap/pywrapgen.py
Comment (jameskermode) [inline f90wrap/pywrapgen.py:159]: default value of `type_check` should probably be `False`, not `None`?

**Review by jameskermode — APPROVED** [review]
Thank you for this contribution. Looks good from a quick scan apart from the one minor comment noted.

I wasn't able to figure out how this allows polymorphism, but I'll read it again more carefully later before merging.

**Comment by jameskermode** [comment]
OK, I understand the point about polymorphism now. Happy to merge once you fix the minor issue with the `type_check` default value.

**Comment by danbeibei** [comment]
I fixed the type_check default value. I also reworked a bit the case where the fortran type is not known to f90wrap.

**Review by jameskermode — APPROVED** [review]

**Pull request merged**
